### ISSUE TITLE
Potential Bug: SyntheticText2SQL evaluation penalizes exact text duplicates as false positives

### ISSUE THREAD

**Issue by lijierui** [embeddings-benchmark/mteb#4626]
### Describe the bug

In mteb/SyntheticText2SQL, the test retrieval corpus pools both train and test partitions. However, over 94% of the test queries and documents have exact text duplicates in the train partition under different IDs.
Because the qrels only map the query to the official test document ID, retrieving the identical train document is incorrectly scored as a false positive. This artificially tanks Top-1 / NDCG metrics.

If a model maps both to the exact same vector, standard tie-breaking (e.g., argsort) often favors the lower-index train ID (c91481). Because c91481 isn't in the qrels, it registers as a false positive despite being the correct SQL answer.

Suggest Fix: To directly delete the test data from the training set to avoid this issue and data leakage.
Might worth running a full check for other datasets as well, especially those whose corpus includes the training data.

### To reproduce

Concrete Example
```
Train ID: c91481 | Test ID: c100010
Doc Text: "What is the average sales quantity and sales amount for each drug..." (Identical for both)

MTEB qrel: q100010 -> c100010 ONLY.
```
Check example on:
https://huggingface.co/datasets/mteb/synthetic-text2sql/viewer/queries?row=100009
https://huggingface.co/datasets/mteb/synthetic-text2sql/viewer/corpus?row=100009
https://huggingface.co/datasets/mteb/synthetic-text2sql/viewer/queries?row=91480
https://huggingface.co/datasets/mteb/synthetic-text2sql/viewer/corpus?row=91480

### Are you interested to contribute a fix for this bug?

Yes

**Comment by Samoed** [comment]
> Suggest Fix: To directly delete the test data from the training set to avoid this issue and data leakage.

I think we could probably split this dataset into 2 different splits. When I was integrating I didn't do that to have same approach as in CoIR

**Comment by lijierui** [comment]
@Samoed Hi Sam, thanks for the reply. After I checked CoIR's dataset, I found that this issue is crucial for at least 3 of the benchmarks here. (other benchmarks <1% overlapped)

With over 94% of the documents having duplicates in the training set, the scores become meaningless: For 94% of the cases, whether the correct id is in rank 1 or rank 2 is a completely random game based on the floating-point precision, or position-encoding bias where identical docs are in different batches.

Task | #queries in test | # corpus in test | #duplicate train docs in test | % overlapped docs
-- | -- | -- | -- | --
SyntheticText2SQL | 5,851 | 105,851 | 5,517 | 94.29%
CosQA | 500 | 20,604 | 424 | 84.80%
CodeFeedbackST | 31,306 | 156,526 | 4,014 | 12.82%

While CoIR team hasn't responded to my issue https://github.com/CoIR-team/coir/issues/26 (I also put the code to reproduce the finding), and current models are already trained on the training set. 
Maybe MTEB can make changes for the above-mentioned benchmarks to make the evaluation fair for existing models.
1. Changing the evaluation from "matching the doc-id" to "matching the doc-text"
2. Masking out duplicate docs from the training set when building up the retrieval corpus
3. Changing the main score from ncdg@10 or other "rank 1st" sensitive scores to recall@2 or recall@3

In addition, if the docs in the training set have lower indexes than their duplicate in the test set, and when similarity scores are identical, training docs will be picked due to the nature of argsort. Currently, I found that models with left padding benefit by ~10% absolute performance on SyntheticText2SQL, as left padding yields a different position encoding, giving more randomness for picking among 2 identical texts.

**Comment by Samoed** [comment]
I think you can reupload a new version of tasks with filter cleaned samples and we can integrate it

**Comment by lijierui** [comment]
> I think you can reupload a new version of tasks with filter cleaned samples and we can integrate it

Hi Roman @Samoed, below is a script to clean out duplicates in these three benchmarks. Written by GPT5.5, It's not long so I personally reviewed the implementation, it looks good to me. The resulted `cleaned_corpus_test` is the same format as `load_dataset(dataset_name, "corpus")["test"]` and same for the queries and qrels as well, ready for `push_to_hub`.
The overall logic was: if any document in the test set has a duplicate, that duplicate was removed. If the duplicate was another entry in the test set, the entry was removed. If that was in the train set, only the train-doc in the corpus was removed.
Please verify this and intergrate if you find the change reasonable:

```py
#!/usr/bin/env python3
"""Clean duplicate test/corpus rows from affected MTEB retrieval datasets."""
from collections import Counter
from datasets import load_dataset

DATASETS_TO_CLEAN = ["mteb/CosQA", "mteb/SyntheticText2SQL", "mteb/CodeFeedbackST"]

def remove_duplicate_test_pairs(queries_test, qrels_test, corpus_test):
    doc_text_by_id = {row["_id"]: row["text"] for row in corpus_test}
    seen_doc_texts = set()
    removed_query_ids = set()
    removed_doc_ids = set()
    kept_doc_ids = set()

    for row in qrels_test:
        query_id = row["query-id"]
        doc_id = row["corpus-id"]
        doc_text = doc_text_by_id[doc_id]
        if doc_text in seen_doc_texts:
            removed_query_ids.add(query_id)
            if doc_id not in kept_doc_ids:
                removed_doc_ids.add(doc_id)
        else:
            seen_doc_texts.add(doc_text)
            kept_doc_ids.add(doc_id)

    cleaned_queries_test = queries_test.filter(lambda row: row["_id"] not in removed_query_ids)
    cleaned_qrels_test = qrels_test.filter(lambda row: row["query-id"] not in removed_query_ids)
    cleaned_corpus_test = corpus_test.filter(lambda row: row["_id"] not in removed_doc_ids)
    return cleaned_queries_test, cleaned_qrels_test, cleaned_corpus_test


def remove_duplicate_corpus_rows(qrels_test, corpus_test):
    test_doc_ids = {row["corpus-id"] for row in qrels_test}
    test_doc_texts = {row["text"] for row in corpus_test if row["_id"] in test_doc_ids}
    cleaned_corpus_test = corpus_test.filter(lambda row: row["_id"] in test_doc_ids or row["text"] not in test_doc_texts)
    return cleaned_corpus_test

def count_duplicate_test_docs_in_corpus(qrels_test, corpus_test) -> int:
    test_doc_ids = {row["corpus-id"] for row in qrels_test}
    test_doc_texts = {row["text"] for row in corpus_test if row["_id"] in test_doc_ids}
    test_doc_text_counts = Counter(row["text"] for row in corpus_test if row["text"] in test_doc_texts)
    return sum(count - 1 for count in test_doc_text_counts.values())

def clean_dataset(dataset_name: str) -> None:
    corpus = load_dataset(dataset_name, "corpus")
    queries = load_dataset(dataset_name, "queries")
    qrels = load_dataset(dataset_name, "qrels")

    original_queries_test = queries["test"]
    original_qrels_test = qrels["test"]
    original_corpus_test = corpus["test"]

    queries_test, qrels_test, corpus_after_test_dedup = remove_duplicate_test_pairs(original_queries_test, original_qrels_test, original_corpus_test)
    cleaned_corpus_test = remove_duplicate_corpus_rows(qrels_test, corpus_after_test_dedup)

    original_test_pairs = len(original_qrels_test)
    original_corpus = len(original_corpus_test)
    removed_test_pairs = original_test_pairs - len(qrels_test)
    removed_corpus_train_entries = len(corpus_after_test_dedup) - len(cleaned_corpus_test)
    still_duplicate_test_doc = count_duplicate_test_docs_in_corpus(qrels_test, cleaned_corpus_test)

    if len(queries_test) != len(qrels_test):
        raise AssertionError("queries/test and qrels/test sizes do not match")
    if still_duplicate_test_doc:
        raise AssertionError(f"{still_duplicate_test_doc} duplicate test docs remain")

    print(f"Handling Dataset {dataset_name}")
    print(f"1. original test pairs: {original_test_pairs}")
    print(f"2. original corpus: {original_corpus}")
    print(f"3. removed test pairs: {removed_test_pairs}")
    print(f"4. removed corpus train entries: {removed_corpus_train_entries}")
    print(f"5. still duplicate test doc: {still_duplicate_test_doc}")
    print("=================================")


def main() -> None:
    for dataset_name in DATASETS_TO_CLEAN:
        clean_dataset(dataset_name)


if __name__ == "__main__":
    main()

```
<!DOCTYPE html>
Dataset | # Original test pairs | # Original corpus | # Removed test pairs | # Removed corpus train docs
-- | -- | -- | -- | --
mteb/CosQA | 500 | 20,604 | 28 | 2,570
mteb/SyntheticText2SQL | 5,851 | 105,851 | 2 | 5,613
mteb/CodeFeedbackST | 31,306 | 156,526 | 524 | 4,263

**Comment by Samoed** [comment]
Looks good. Can you upload updated tasks and run some models to see performance?

**Comment by Samoed** [comment]
Yes, that would be great
